osquery/tables/system/darwin/mdls.mm

```diff
@@ -0,0 +1,129 @@
+/**
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under both the Apache 2.0 license (found in the
+ *  LICENSE file in the root directory of this source tree) and the GPLv2 (found
+ *  in the COPYING file in the root directory of this source tree).
+ *  You may select, at your option, one of the above-listed licenses.
+ */
+
+#include <CoreServices/CoreServices.h>
+
+#include <boost/filesystem.hpp>
+#include <osquery/core.h>
+#include <osquery/filesystem/filesystem.h>
+#include <osquery/logger.h>
+#include <osquery/tables.h>
+#include <osquery/utils/conversions/darwin/cfdata.h>
+#include <osquery/utils/conversions/darwin/cfnumber.h>
+#include <osquery/utils/conversions/darwin/cfstring.h>
+
+namespace osquery {
+namespace tables {
+
+const int MAX_RECURSION_DEPTH = 2;
+
+std::string getStringOfValue(CFTypeRef value, int depth) {
+  if (depth >= MAX_RECURSION_DEPTH) {
+    return "";
+  }
+  std::string rvalue;
+
+  if (CFGetTypeID(value) == CFNumberGetTypeID()) {
+    rvalue = stringFromCFNumber(static_cast<CFDataRef>(value));
+  } else if (CFGetTypeID(value) == CFDateGetTypeID()) {
+    auto unix_time = CFDateGetAbsoluteTime(static_cast<CFDateRef>(value)) +
+                     kCFAbsoluteTimeIntervalSince1970;
+    rvalue = INTEGER(std::llround(unix_time));
+  } else if (CFGetTypeID(value) == CFArrayGetTypeID()) {
+    for (auto i = 0; i < CFArrayGetCount(static_cast<CFArrayRef>(value)); ++i) {
+      CFTypeRef b = CFArrayGetValueAtIndex(static_cast<CFArrayRef>(value), i);
+      if (i == 0) {
+        rvalue = getStringOfValue(b, depth + 1);
+      } else {
+        // beware of recursion....
+        rvalue = rvalue + "," + getStringOfValue(b, depth + 1);
+      }
+    }
+  } else if (CFGetTypeID(value) == CFBooleanGetTypeID()) {
+    if (CFBooleanGetValue(static_cast<CFBooleanRef>(value))) {
+      rvalue = "true";
+    } else {
+      rvalue = "false";
+    }
+  } else if (CFGetTypeID(value) == CFStringGetTypeID()) {
+    rvalue = stringFromCFString(static_cast<CFStringRef>(value));
+  } else {
+    rvalue = "null";
+  }
+
+  return rvalue;
+}
+void genResults(const std::string& path, QueryData& results) {
+  CFStringRef cs = CFStringCreateWithCString(
+      kCFAllocatorDefault, path.c_str(), kCFStringEncodingASCII);
+  MDItemRef mdi = MDItemCreate(kCFAllocatorDefault, cs);
```

Comment:
Does this need to be released? Not clear based on the docs: https://developer.apple.com/documentation/coreservices/1426917-mditemcreate?language=objc

Reply:
I don't see any reason why we shouldn't release it.